OpenRA.Mods.Cnc/Traits/ResourcePurifier.cs

```diff
@@ -0,0 +1,90 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Gives additional cash when resources are delivered to refineries.")]
+	public class ResourcePurifierInfo : ConditionalTraitInfo
+	{
+		[FieldLoader.Require]
+		[Desc("Percentage value of the resource to grant as cash.")]
+		public readonly int Modifier = 0;
+
+		[Desc("Whether to show the cash tick indicators rising from the actor.")]
+		public readonly bool ShowTicks = true;
+
+		[Desc("How long the cash ticks stay on the screen.")]
+		public readonly int TickLifetime = 30;
+
+		[Desc("How often the cash ticks can appear.")]
+		public readonly int TickRate = 10;
+
+		public override object Create(ActorInitializer init) { return new ResourcePurifier(init.Self, this); }
+	}
+
+	public class ResourcePurifier : ConditionalTrait<ResourcePurifierInfo>, INotifyResourceDumped, ITick, INotifyOwnerChanged
+	{
+		readonly int[] modifier;
+
+		PlayerResources playerResources;
```

Comment:
`readonly`

Reply:
This is being set in INotifyOwnerChanged.OwnerChanged.